core/src/main/scala/org/bykn/bosatsu/SourceConverter.scala

```diff
@@ -1011,6 +1133,105 @@ object SourceConverter {
     localDefs: Stream[TypeDefinitionStatement]): SourceConverter =
     new SourceConverter(thisPackage, imports, localDefs)
 
+  private def concat[A](ls: List[A], tail: NonEmptyList[A]): NonEmptyList[A] =
+    ls match {
+      case Nil => tail
+      case h :: t => NonEmptyList(h, t ::: tail.toList)
+    }
+
+  /**
+   * For all duplicate binds, for all but the final
+   * value, rename them
+   */
+  def makeLetsUnique[D](
+    lets: List[(Bindable, RecursionKind, D)])(
+    newName: (Bindable, Int) => (Bindable, D => D)): List[(Bindable, RecursionKind, D)] =
+      NonEmptyList.fromList(lets) match {
+        case None => Nil
+        case Some(nelets) =>
+          // there is at least 1 let, but maybe no duplicates
+          val dups: Map[Bindable, Int] =
+            nelets.foldLeft(Map.empty[Bindable, Int]) {
+              case (bound, (b, _, _)) =>
+                bound.get(b) match {
+                  case Some(c) => bound.updated(b, c + 1)
+                  case None => bound.updated(b, 1)
+                }
+            }
+            .filter { case (_, v) => v > 1 }
+
+          if (dups.isEmpty) {
+            // no duplicated top level names
+            lets
+          }
+          else {
+            // we rename all but the last name for each duplicate
+            type BRD = (Bindable, RecursionKind, D)
+
+            @annotation.tailrec
+            def renameUntilNext(name: Bindable, lets: NonEmptyList[BRD], acc: List[BRD])(fn: D => D): NonEmptyList[BRD] = {
+              val NonEmptyList(head @ (b, r, d), tail) = lets
+              if (b == name) {
+                val head1 =
+                  if (r.isRecursive) {
+                    // the new b is in scope right away
+                    head
+                  }
+                  else {
+                    // the old b1 is in scope for this one
+                    (b, r, fn(d))
+                  }
+                NonEmptyList(head1, acc).reverse.concat(tail)
+              }
+              else {
+                NonEmptyList.fromList(tail) match {
+                  case Some(netail) =>
+                    // this b is different from name, but may reference it
+                    val d1 = fn(d)
+                    renameUntilNext(name, netail, (b, r, d1) :: acc)(fn)
+                  case None =>
+                    // head is the last, and the last is never renamed
+                    NonEmptyList(head, acc).reverse
```

Comment:
This branch can't happen because it means we are renaming but we never hit the bound name. That implies the name isn't actually duplicated. We should throw here that an invariant has been violated.

Note the code coverage does not reach here.